File: sessionport/store.py
import os
import re
from abc import ABC, abstractmethod
from pathlib import Path
from typing import TYPE_CHECKING
from urllib.parse import unquote, urlparse

if TYPE_CHECKING:
    from mypy_boto3_s3.client import S3Client

DEFAULT_STORE = os.path.join(str(Path.home()), ".sessionport", "store")


class Store(ABC):
    @abstractmethod
    def describe(self) -> str: ...

    @abstractmethod
    def put(self, key: str, data: bytes) -> None: ...

    @abstractmethod
    def get(self, key: str) -> bytes: ...


class FileStore(Store):
    def __init__(self, root: str) -> None:
        self._root = root

    def describe(self) -> str:
        return self._root

    def put(self, key: str, data: bytes) -> None:
        path = Path(self._root) / key
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_bytes(data)

    def get(self, key: str) -> bytes:
        return (Path(self._root) / key).read_bytes()


class S3Store(Store):
    def __init__(self, bucket: str, prefix: str) -> None:
        self._bucket = bucket
        self._prefix = prefix

    def describe(self) -> str:
        return f"s3://{self._bucket}/{self._prefix}"

    def _key_for(self, key: str) -> str:
        return f"{self._prefix}/{key}" if self._prefix else key

    # Imported lazily so the file backend never pays for loading the SDK.
    def _client(self) -> "S3Client":
        import boto3
        from botocore.config import Config

        endpoint = os.environ.get("SESSIONPORT_S3_ENDPOINT")
        region = os.environ.get("SESSIONPORT_S3_REGION")
        return boto3.client(
            "s3",
            endpoint_url=endpoint,
            region_name=region,
            # A custom endpoint means MinIO/R2/Garage, which need path style.
            config=Config(s3={"addressing_style": "path"}) if endpoint else None,
        )

    def put(self, key: str, data: bytes) -> None:
        self._client().put_object(
            Bucket=self._bucket,
            Key=self._key_for(key),
            Body=data,
            ContentType="application/gzip",
        )

    def get(self, key: str) -> bytes:
        response = self._client().get_object(
            Bucket=self._bucket, Key=self._key_for(key)
        )
        body = response.get("Body")
        if body is None:
            raise ValueError(f"empty object at {self._key_for(key)}")
        data: bytes = body.read()
        return data


def create_store(uri: str) -> Store:
    if uri.startswith("s3://"):
        without_scheme = uri[len("s3://") :]
        slash = without_scheme.find("/")
        bucket = without_scheme if slash == -1 else without_scheme[:slash]
        prefix = "" if slash == -1 else re.sub(r"/$", "", without_scheme[slash + 1 :])
        if not bucket:
            raise ValueError(f"invalid S3 store URI: {uri}")
        return S3Store(bucket, prefix)

    if uri.startswith("file://"):
        return FileStore(unquote(urlparse(uri).path))
    return FileStore(uri if os.path.isabs(uri) else os.path.abspath(uri))


def resolve_store_uri(explicit: str | None = None) -> str:
    if explicit is not None:
        return explicit
    return os.environ.get("SESSIONPORT_STORE", DEFAULT_STORE)

File: src/store.ts
import { mkdirSync, readFileSync, writeFileSync } from 'node:fs'
import { homedir } from 'node:os'
import { dirname, isAbsolute, join, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

export interface Store {
  describe(): string
  put(key: string, data: Buffer): Promise<void>
  get(key: string): Promise<Buffer>
}

export const DEFAULT_STORE = join(homedir(), '.sessionport', 'store')

class FileStore implements Store {
  constructor(private readonly root: string) {}

  describe(): string {
    return this.root
  }

  async put(key: string, data: Buffer): Promise<void> {
    const path = join(this.root, key)
    mkdirSync(dirname(path), { recursive: true })
    writeFileSync(path, data)
  }

  async get(key: string): Promise<Buffer> {
    return readFileSync(join(this.root, key))
  }
}

class S3Store implements Store {
  constructor(
    private readonly bucket: string,
    private readonly prefix: string
  ) {}

  describe(): string {
    return `s3://${this.bucket}/${this.prefix}`
  }

  private keyFor(key: string): string {
    return this.prefix ? `${this.prefix}/${key}` : key
  }

  // Imported lazily so the file backend never pays for loading the SDK.
  private async sdk() {
    const sdk = await import('@aws-sdk/client-s3')
    const endpoint = process.env['SESSIONPORT_S3_ENDPOINT']
    const region = process.env['SESSIONPORT_S3_REGION']
    const client = new sdk.S3Client({
      ...(endpoint ? { endpoint, forcePathStyle: true } : {}),
      ...(region ? { region } : {})
    })
    return { sdk, client }
  }

  async put(key: string, data: Buffer): Promise<void> {
    const { sdk, client } = await this.sdk()
    await client.send(
      new sdk.PutObjectCommand({
        Bucket: this.bucket,
        Key: this.keyFor(key),
        Body: data,
        ContentType: 'application/gzip'
      })
    )
  }

  async get(key: string): Promise<Buffer> {
    const { sdk, client } = await this.sdk()
    const response = await client.send(
      new sdk.GetObjectCommand({ Bucket: this.bucket, Key: this.keyFor(key) })
    )
    if (!response.Body) throw new Error(`empty object at ${this.keyFor(key)}`)
    return Buffer.from(await response.Body.transformToByteArray())
  }
}

export function createStore(uri: string): Store {
  if (uri.startsWith('s3://')) {
    const withoutScheme = uri.slice('s3://'.length)
    const slash = withoutScheme.indexOf('/')
    const bucket = slash === -1 ? withoutScheme : withoutScheme.slice(0, slash)
    const prefix = slash === -1 ? '' : withoutScheme.slice(slash + 1).replace(/\/$/, '')
    if (!bucket) throw new Error(`invalid S3 store URI: ${uri}`)
    return new S3Store(bucket, prefix)
  }

  if (uri.startsWith('file://')) return new FileStore(fileURLToPath(uri))
  return new FileStore(isAbsolute(uri) ? uri : resolve(uri))
}

export function resolveStoreUri(explicit?: string): string {
  return explicit ?? process.env['SESSIONPORT_STORE'] ?? DEFAULT_STORE
}
